ls1intum/Artemis
# Enhancement/Programming Exercise/Improve failed submission handling

### Checklist

### Motivation and Context



Improve the failed submission handling for programming exercises: https://github.com/ls1intum/Artemis/issues/991. 
See https://github.com/ls1intum/Artemis/pull/707 for generell information on the feature.

### Description


Instead of always just retriggering a failed build, we go through the following checks/steps:

1) We query Bamboo for the last build result of the given build plan. If it exists and matches the commitHash of the last commit, we just save this result.

2) We check if any build is running for the participation. If so, we inform the client so it can show the 'is building' label on the result comp. 
Note: I did not implement any further polling logic as suggested, as this would need more time.
Note 2: I did not see an easy way to check if the running build belongs to the latestCommit.

3) The last fallback is to just retrigger the build on Bamboo.

### Steps for Testing


**Scenario 1: Failed submission (CIS executed build but did/could not inform Artemis about created build result)**

1. Log in to Artemis
2. Go to the code-editor, submit
3. Delete the created result + feedbacks from your db
4. Wait 2 minutes => Artemis will now consider the submission failed
5. Click on the red retry button next to the result component
=> Artemis should not trigger the build again, but try to retrieve the build results from Bamboo.

**Scenario 2: Failed submission (CIS has high build activity, build is either queued or running)**

1. Log in to Artemis
2. Go to the code-editor, submit
3. Cancel the build in Bamboo
4. Wait 2 minutes => Artemis will now consider the submission failed
5. Trigger the build on Bamboo, quickly click on the red retry button next to the result component
=> Artemis should not trigger the build again, but find out that the build is still running on Bamboo

### Screenshots
